FirebasePhoneAuthUI/CountryCode/FUIPhoneNumber.m

```diff
@@ -0,0 +1,111 @@
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+#import "FUIPhoneNumber.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+NSString * const FUIPhoneNumberValidationErrorDomain = @"FUIPhoneNumberValidationErrorDomain";
+
+@implementation FUIPhoneNumber
+
+- (instancetype)initWithNormalizedPhoneNumber:(NSString *)normalizedPhoneNumber {
+  if (!normalizedPhoneNumber) {
+    return nil;
+  }
+  NSString *rawPhoneNumber;
+  FUICountryCodes *codes = [[FUICountryCodes alloc] init];
+  FUICountryCodeInfo *countryCode = [codes countryCodeInfoForPhoneNumber:normalizedPhoneNumber];
+
+  if (countryCode) {
+    // Add 1 for the '+' character
+    NSInteger countryCodeLength = countryCode.dialCode.length + 1;
+    if (normalizedPhoneNumber.length >= countryCodeLength) {
+      rawPhoneNumber = [normalizedPhoneNumber substringFromIndex:countryCodeLength];
+    }
+  }
+  if (!rawPhoneNumber) {
+    rawPhoneNumber = normalizedPhoneNumber;
+    countryCode = [codes countryCodeInfoFromDeviceLocale];
+  }
+  return [self initWithRawPhoneNumber:rawPhoneNumber countryCode:countryCode];
+}
+
+- (instancetype)initWithNormalizedPhoneNumber:(NSString *)normalizedPhoneNumber
+                               rawPhoneNumber:(NSString *)rawPhoneNumber
+                                  countryCode:(FUICountryCodeInfo *)countryCode {
+  NSParameterAssert(normalizedPhoneNumber);
+  NSParameterAssert(rawPhoneNumber);
+  NSParameterAssert(countryCode);
+  if (!normalizedPhoneNumber || !rawPhoneNumber || !countryCode){
+    return nil;
+  }
+  if (self = [super init]) {
+    _countryCode = countryCode;
+    _rawPhoneNumber = rawPhoneNumber;
+    _normalizedPhoneNumber = normalizedPhoneNumber;
+  }
+  return self;
+}
+
+- (instancetype)initWithRawPhoneNumber:(NSString *)rawPhoneNumber
+                           countryCode:(FUICountryCodeInfo *)countryCode {
+  NSString *dialCode = countryCode.dialCode;
+  if ([dialCode characterAtIndex:0] != '+') {
+    dialCode = [@"+" stringByAppendingString:dialCode];
+  }
+  NSString *normalizedPhoneNumber = [NSString stringWithFormat:@"%@%@", dialCode, rawPhoneNumber];
+  
+  return [self initWithNormalizedPhoneNumber:normalizedPhoneNumber
+                              rawPhoneNumber:rawPhoneNumber
+                                 countryCode:countryCode];
+}
+
+- (BOOL)validate:(NSError *__autoreleasing *)error {
+  // The first character is always the '+'
+  BOOL firstCharacterIsPlus = [self.normalizedPhoneNumber characterAtIndex:0] == '+';
+  if (!firstCharacterIsPlus) {
+    if (error) {
+      *error = [NSError errorWithDomain:FUIPhoneNumberValidationErrorDomain
+                                   code:FUIPhoneNumberValidationErrorMissingPlus
+                               userInfo:nil];
```

Comment:
We should be adding userInfo dictionaries with description keys, so meaningful error messages appear when the errors are logged.